alicloud/resource_alicloud_datahub_topic.go

```diff
@@ -0,0 +1,216 @@
+		Lifecycle:   lifeCycle,
+		Comment:     topicComment,
+	}
+
+	if recordType == "TUPLE" {
+		t.RecordType = datahub.TUPLE
+		t.RecordSchema = getRecordSchema(recordSchema)
+	} else if recordType == "BLOB" {
+		t.RecordType = datahub.BLOB
+	} else {
+		return fmt.Errorf("failed to create topic'%s/%s' with invalid record type: %s", projectName, topicName, recordType)
+	}
+
+	err := dh.CreateTopic(t)
+	if err != nil {
+		return fmt.Errorf("failed to create topic'%s/%s' with error: %s", projectName, topicName, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s%s%s", projectName, COLON_SEPARATED, topicName))
+	return resourceAliyunDatahubTopicRead(d, meta)
+}
+
+func parseId2(d *schema.ResourceData, meta interface{}) (projectName, topicName string, err error) {
+	split := strings.Split(d.Id(), COLON_SEPARATED)
+	if len(split) != 2 {
+		err = fmt.Errorf("you should use resource alicloud_datahub_topic's new field 'project_name' and 'name' to re-import this resource.")
+		return
+	} else {
+		projectName = split[0]
+		topicName = split[1]
+		return
+	}
+}
+
+func resourceAliyunDatahubTopicRead(d *schema.ResourceData, meta interface{}) error {
+	projectName, topicName, err := parseId2(d, meta)
+	if err != nil {
+		return err
+	}
+
+	dh := meta.(*AliyunClient).dhconn
+
+	topic, err := dh.GetTopic(projectName, topicName)
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+		}
+		return fmt.Errorf("failed to access topic '%s/%s' with error: %s", projectName, topicName, err)
+	}
+
+	d.Set("name", topic.TopicName)
+	d.Set("project_name", topic.ProjectName)
+	d.Set("shard_count", topic.ShardCount)
+	d.Set("life_cycle", topic.Lifecycle)
+	d.Set("comment", topic.Comment)
+	d.Set("record_type", topic.RecordType.String())
+	d.Set("record_schema", topic.RecordSchema.String())
+	d.Set("create_time", datahub.Uint64ToTimeString(topic.CreateTime))
+	d.Set("last_modify_time", datahub.Uint64ToTimeString(topic.LastModifyTime))
+	return nil
+}
+
+func resourceAliyunDatahubTopicUpdate(d *schema.ResourceData, meta interface{}) error {
+	projectName, topicName, err := parseId2(d, meta)
+	if err != nil {
+		return err
+	}
+
+	dh := meta.(*AliyunClient).dhconn
+
+	if !d.IsNewResource() && (d.HasChange("life_cycle") || d.HasChange("comment")) {
+		lifeCycle := d.Get("life_cycle").(int)
+		topicComment := d.Get("comment").(string)
+
+		err = dh.UpdateTopic(projectName, topicName, lifeCycle, topicComment)
+		if err != nil {
+			return fmt.Errorf("failed to update topic '%s/%s' with error: %s", projectName, topicName, err)
+		}
+	}
+
+	return resourceAliyunDatahubTopicRead(d, meta)
+}
+
+func resourceAliyunDatahubTopicDelete(d *schema.ResourceData, meta interface{}) error {
+	projectName, topicName, err := parseId2(d, meta)
+	if err != nil {
+		return err
+	}
+
+	dh := meta.(*AliyunClient).dhconn
+
+	return resource.Retry(3*time.Minute, func() *resource.RetryError {
+		_, err := dh.GetTopic(projectName, topicName)
+
+		if err != nil {
+			if isRetryableDatahubError(err) {
```

Comment:
I the error is not found error, please return nil directly.

Reply:
done